codesearch/token/token.go

```diff
@@ -362,11 +365,20 @@ func (tt *SparseNgramTokenizer) refillLine() error {
 		if err := tt.scanner.Err(); err != nil {
 			return err
 		}
-		loweredLine := strings.ToLower(tt.scanner.Text())
-		if len(loweredLine) == 0 {
+		buf := tt.scanner.Bytes()
+		lineLength := len(buf)
+
+		if lineLength == 0 {
 			continue
 		}
-		tt.s = []rune(loweredLine)
+		tt.s = tt.s[:lineLength]
```

Comment:
This is weird, right? You're indexing into runes with a byte offset? I assume I'm misreading but figure it's faster to ask.

Reply:
it is weird! I actually had to google this. The gist of it is that `tt.s` is a rune buffer with sufficient capacity to hold any line we'd read. We scan a line of bytes from the scanner and ensure the length of `tt.s` is at least that long so that we have room to put all decoded runes in there (rune count being <= byte count, depending on encoding).

Once we've decoded all runes and know the exact length, we set `tt.s`'s length to the number of runes (capacity is unchanged, and it's the same memory).